src/azul/service/index_query_service.py

```diff
@@ -70,38 +68,40 @@ def get_data(self,
                                           filters=filters,
                                           pagination=pagination,
                                           entity_type=entity_type)
-        # FIXME: Generalize file URL injection.
-        #        https://github.com/DataBiosphere/azul/issues/2545
-        if entity_type in ('files', 'bundles'):
-            # Inject URLs to data files
-            for hit in response['hits']:
-                for file in hit['files']:
-                    file['url'] = file_url_func(catalog=catalog,
-                                                file_uuid=file['uuid'],
-                                                version=file['version'])
-        elif entity_type == 'projects':
-            # Similarly, inject URLs to matrix files in stratification trees
-            def transform(node: Union[JSONs, JSON]) -> Union[MutableJSONs, MutableJSON]:
-                if isinstance(node, dict):
-                    return {k: transform(v) for k, v in node.items()}
-                elif isinstance(node, list):
-                    return [
-                        {
-                            'name': file['name'],
-                            'source': file['source'],
-                            'url': file_url_func(catalog=catalog,
-                                                 file_uuid=file['uuid'],
-                                                 version=file['version'])
-                        }
-                        for file in node
-                    ]
-                else:
-                    assert False
 
-            for hit in response['hits']:
-                for project in hit['projects']:
-                    for key in 'matrices', 'contributorMatrices':
-                        project[key] = transform(project[key])
+        def inject_file_urls(node: AnyMutableJSON, *path: str) -> None:
+            if node is None:
+                pass
+            elif isinstance(node, (str, int, float, bool)):
+                pass
+            elif isinstance(node, list):
+                for child in node:
+                    inject_file_urls(child, *path)
+            elif isinstance(node, dict):
+                if path:
+                    try:
+                        inject_file_urls(node[path[0]], *path[1:])
+                    except KeyError:
+                        pass
```

Comment:
Would it be useful to have contextual information about the current node and path?

Reply:
A `KeyError` here means the specified path doesn't exist in the given node tree. Unfortunately this can't be a failure however due to the generic nature of the function which will process responses from many different indices ('projects', 'files', 'samples') and a path such as ('hits', 'projects', 'matrices') won't exist in the 'files' index response.
I think the `pass` does deserve a short comment however. I will add that now.